libnfc/drivers/pcsc.c

```diff
@@ -184,6 +193,118 @@ uint8_t pcsc_get_icc_type(struct nfc_device *pnd)
   return it;
 }
 
+char* pcsc_get_vendor_name(struct nfc_device *pnd){
+  struct pcsc_data *data = pnd->driver_data;
+  LPBYTE vendor_name = NULL;
+  DWORD vendor_name_len = SCARD_AUTOALLOCATE;
+
+  int res = SCardGetAttrib(data->hCard, SCARD_ATTR_VENDOR_NAME, (LPBYTE)&vendor_name, &vendor_name_len);
+  if (!res && vendor_name && vendor_name_len > 0 && vendor_name[0] != '\0') {
+    char *psVendorName = (char *)malloc(sizeof(char) * vendor_name_len);
+    memcpy(psVendorName, vendor_name, vendor_name_len);
+    return psVendorName;
+  }
+  return NULL;
+}
+
+bool is_pcsc_reader_vendor(struct nfc_device *pnd, const char * target_vendor_name)
+{
+  bool isTarget = false;
+  char * sReaderVendorName = pcsc_get_vendor_name(pnd);
+  if (sReaderVendorName)
+  {
+    isTarget = (strstr(sReaderVendorName, target_vendor_name)) ? true:false;
+    free(sReaderVendorName);
+    sReaderVendorName = NULL;
+  }
+
+  return isTarget;
+}
+
+bool is_pcsc_reader_vendor_feitian(struct nfc_device *pnd)
+{
+  return is_pcsc_reader_vendor(pnd, "Feitian") || is_pcsc_reader_vendor(pnd, "FeiTian") ||is_pcsc_reader_vendor(pnd, "feitian");
```

Comment:
could we avoid calling three times pcsc_get_vendor_name and interrogating three times the device ?

Reply:
given that is_pcsc_reader_vendor_feitian is called at multiple places, best is probably to memorize the PCSC vendor name in the driver once it's known

Reply:
Yes, make sense, will do modify the code, thanks for suggestions.